### ISSUE TITLE
Knife and Incendiary Grenade OP above 60fps 

### ISSUE THREAD

**Issue by DarkxVoid-loop** [nipkownix/re4_tweaks#263]
### Describe the feature you'd like to see added

When you release the re4 fps and be able to play with 90, 120, 150.... fps, the Incendiary grenade, flash and knife will be very OP
because their damage is per tick and the flash stun time is also per tick, so it would have to be fixed to make the game balanced

**Comment by linkthehylian** [comment]
Being able to go above 60 fps is only available in certain versions of re4_tweaks and hasn't been publicly released yet. Once it does though, I think that will be the time to look into it.

> What is a tick? What is OP?

I think they meant, "per second" since the term "tick" is mainly used when referring to online games.
Also "OP" is an acronym for "overpowered".

**Comment by emoose** [comment]
Tick is also used in offline games too, mostly known from source engine games, but pretty sure other games use a similar thing too.

It's pretty much just how many times a second the game actually updates objects etc, similar to framerate but usually lower so less work needs to be done, eg. game might present at 120Hz but only actually update positions etc 30 times a second, so tickrate is 30.

In RE4s case the tickrate seems to be tied to the framerate, not sure if it's updating at exactly the same as framerate or if [it's 1/2 of framerate](https://github.com/nipkownix/re4_tweaks/issues/23#issuecomment-984812169) though (the 1/2 framerate stuff might just be used for a few weird things)

About the OP, I guess that probably means the damage is also increased at 60FPS too? Maybe worth looking into some time, I haven't really looked into how game handles damage yet though (something to do with `YARARE_INFO` maybe), if anyone knows where the values are located in the EXE maybe I can trace where it's used from there.

**Comment by DarkxVoid-loop** [comment]
yes the damage of certain weapons and grenades is linked to the game frame, at 30fps Mendez dies with 6 Incendiary grenades, the flash grenade lasts 1 minute and the ganados do not hit axes and arrows 
already at 60 fps it's 4 grenades to kill Mendez, the flash grenade lasts almost 2 minutes and so on, I don't know why it's linked to the frame

**Comment by emoose** [comment]
Seems `GetWepDmVal` is used to get damage amount for most weapons, except some other ones are handled differently, most EmXX classes contains a `emXXSetDmVal` which either calls the GetWep func or uses a hardcoded damage amount for certain weapons, that then returns the final damage amount to the `emXXDmCk` func which applies the damage via `LifeDownSet`.

My first guess would be to change `LifeDownSet` so that damage amount given is multiplied by deltaTime, at 60FPS this would halve the damage each tick, which should result in same damage between 30/60 (since 60FPS = double the number of ticks, halving damage amount per tick should fix it)
If there's any weapons which only apply damage in a single tick this would pretty much halve its entire damage though, good chance there's probably some weapons like that here (especially if it's only certain weps that seem to have increased dmg...)

Probably better to figure out which weapons are most affected and then hook GetWepDmVal to do the deltaTime stuff.

**Comment by LethalPlacebo** [comment]
Hey guys, not sure if this helps but Japanese YouTube user wairokku has several videos documenting the differences/bugs between 30FPS and 60FPS on the Steam version of RE4. Here's one of them that shows the incendiary grenade difference that DarkxVoid mentioned:

https://www.youtube.com/watch?v=2Wiq5StTKwE

The other videos are very easy to find on the user's channel, they haven't uploaded much else. It's worth noting that some things listed in the videos, like the boat machine gun in the Separate Ways campaign, were actually addressed in official Capcom patches, but many were not, such as the incendiary grenade or falling item speeds.

Regarding the knife, you can tell that sometimes the knife double slashes on 60FPS, for example when you break a box with a snake in it by slashing it, often times you'll end up automatically killing the snake without having to slash twice.

**Comment by DarkxVoid-loop** [comment]
the knife also at 60fps gives 9 hits on the padlock and at 30fps it only gives one, and the flash grenade stun time lasts longer at 60 fps, but instead of basing it on 30 fps there's no way to leave it at the level of 60 fps

**Comment by LethalPlacebo** [comment]
> the knife also at 60fps gives 9 hits on the padlock and at 30fps it only gives one, and the flash grenade stun time lasts longer at 60 fps, but instead of basing it on 30 fps there's no way to leave it at the level of 60 fps

I'm not sure if there's one for the flashbang, but this video shows the difference when slashing the padlock in the cage:

https://youtu.be/GWNgxjYSw9g?t=162

I did my own testing on this on Professional mode using Area Jump in debug mode. At 30 FPS, the lock takes 4 knife slashes, whereas on 60 FPS it only takes 3 so it seems that this may have been partially patched? Unless RE4 Tweaks is doing something here.

Other thing of note: there are two locks in that cage. The other lock requires 5 knife slashes at 30 FPS, but still only 3 at 60 FPS. Hope any of this info was useful.
